### ISSUE TITLE
Allow ParamConverters to be defined by @Provider

### ISSUE THREAD

**Issue by sdaschner** [eclipse-ee4j/jaxrs-api#878]
One gotcha I've bumped into many times is that you can't register a `ParamConverter` with `@Provider` but only via `ParamConverterProvider`.

While we need to keep the latter one, is there anything holding us back from directly allowing the following, without any additional config?

```java
@Provider
public class LocalDateConverter implements ParamConverter<LocalDate> {
    
    @Override
    public LocalDate fromString(String value) {
        // ...
    }

    @Override
    public String toString(LocalDate value) {
        // ...
    }
}

```

**Comment by mkarg** [comment]
> How's an implementation going to find the _correct_ `ParamConverter` after type erasure?

We could either mark the *correct* converter with an annotation having its type as an attribute, or we could extend `ParamConverter` with a method which allows to ask `boolean canConvert(Type)`.

**Comment by chkal** [comment]
> We could either mark the _correct_ converter with an annotation having its type as an attribute

I don't think that this would work well for generic types.

> or we could extend ParamConverter with a method which allows to ask boolean canConvert(Type)

But that's essentially what `ParamConverterProvider` provides today.

My 2 cents: It is clear that directly annotating `ParamConverter` with `@Provider` won't work in all cases. But it could be a nice shortcut in simple cases (non-generic types without the need to access annotations). For all other cases `ParamConverterProvider` should still be the way to go.

**Comment by sdaschner** [comment]
Good point @spericas, having a solution that works also with generics would be preferable, but that indeed could be solved with a method similar to what @mkarg described. This could be a `default` nop method for when the type is clear (also to be binary compatible), and would need to be extended if the users define multiple converters that only differ in their nested types.
